docs/src/user-guide/tutorials/auction.md

```diff
@@ -0,0 +1,349 @@
+
+```rust
+struct AuctionAlreadyEnded {
+}
+```
+
+The next check is whether the incoming bid exceeds the current highest bid. If not, the bid has failed and it may as well revert. We can repeat the same logic as for `AuctionAlreadyEnded`. We can also report the current highest bid in the revert message to help the user reprice if they want to. Add the following to `auction.fe`:
+
+```rust
+struct BidNotHighEnough {
+    pub highest_bid: u256
+}
+```
+
+> Notice that the value being checked is `msg.value` which is included in the `ctx` object. `ctx` is where you can access incoming transaction data.
+
+Next, if the incoming transaction *is* the highest bid, you need to track how much the sender should receive as a payout if their bid ends up being exceeded by another user (i.e. if they get outbid, they get their ETH back). To do this, you add a key-value pair to the `pending_returns` mapping, with the user address as the key and the transaction amount as the value. Both of these come from `ctx` in the form of `msg.sender` and `msg.value`.
+
+Finally, if the incoming bid *is* the highest, you can emit an event. Events are useful because they provide a cheap way to return data from a  contract as they use logs instead of contract storage. Unlike other smart contract languages, there is no `emit` keyword or `Event` type. Instead, you trigger an event by calling the `emit` method on the `ctx` object. You can pass this method a struct that defines the emitted message. You can add the following struct for this event:
+
+```rust
+struct HighestBidIncreased {
+    #indexed
+    pub bidder: address
+    pub amount: u256
+}
+```
+
+You have now implemented all the logic to handle a bid!
+
+### Withdrawing
+
+A previous high-bidder will want to retrieve their ETH from the contract so they can either walk away or bid again. You therefore need to create a `withdraw` method that the user can call. The function will lookup the user address in `pending_returns`. If there is a non-zero value associated with the user's address, the contract should send that amount back to the sender's address. It is important to first update the value in `pending_returns` and *then* send the ETH to the user, otherwise you are exposing a [re-entrancy](https://www.certik.com/resources/blog/3K7ZUAKpOr1GW75J2i0VHh-what-is-a-reentracy-attack) vulnerability (where a user can repeatedly call the contract and receive the ETH multiple times).
+
+Add the following to the contract to implement the `withdraw` method:
+
+```rust
+pub fn withdraw(mut self, mut ctx: Context) -> bool {
```

Comment:
Just because I haven't seen it mentioned here but the reason why we need to use `mut` with the Context here is because `send_value` is causing changes to the blockchain. If you were just to use something like `ctx.msg_sender()` then you would not need `mut` because we are just accessing information but not causing any changes to the blockchain. Looking at it this way, we have 9 different categories that functions can be categorized in from the perspective of their power (See "ABI conformity" here: https://github.com/ethereum/fe/issues/558)